### PR TITLE
Cleaning up of snapshots and events before snapshots based on time and a # of snapshots

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

File: core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala [hunk 1]
```diff
@@ -75,7 +75,7 @@ object CassandraSpec {
 
   val fallbackConfig = ConfigFactory.parseString(s"""
       akka.loggers = ["akka.persistence.cassandra.SilenceAllTestEventListener"]
-      akka.loglevel = INFO
+      akka.loglevel = DEBUG
```
Comment (johanandren) [inline core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala:78]: Left at debug on purpose?

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala [hunk 2]
```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
...
+   * The returned snapshot metadata can be used to issue deletes for events older than the oldest
+   * snapshot.
+   *
+   * @return the snapshot meta of the oldest remaining snapshot. None if there are no snapshots
+   *
+   */
+  def deleteBeforeSnapshot(
+      persistenceId: String,
+      snapshotsToKeep: Int,
+      keepAfter: Long): Future[Option[SnapshotMetadata]] = {
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:118]: Hard to parse what it does from name/docs but I'm not sure I can come up with anything better...

`deleteSnapshotsBefore(persistenceId, unixTimestamp, nrSnapshotsToKeep)`?

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala [hunk 3]
```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
...
+            if (row.getLong("timestamp") > keepAfter) {
+              count += 1
+              true
+            } else if (count < snapshotsToKeep) {
+              count += 1
+              true
+            } else {
+              false
+            }
+          }
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:136]: Could be cleaner with `.zipWithIndex().takeWhile()`

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala [hunk 4]
```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
...
+  }
+
+  /**
+   * Delete all events before a sequenceNr for the given persistence id.
+   * @param persistenceId the persistence id to delete for
+   * @param toSequenceNr sequence nr (inclusive) to delete up to
+   */
+  def deleteTo(persistenceId: String, toSequenceNr: Long): Future[Done] = {
+    (journal ? CassandraJournal.DeleteEventsTo(persistenceId, toSequenceNr)).mapTo[Done]
+  }
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:164]: If the Cassandra delete returns a number for how many was deleted that would be nice to return here, for logging etc.

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala [hunk 5]
```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
...
+    }
+    snapshots.flatMap(rows => issueSnapshotDelete(persistenceId, maxSnapshotsToKeep, rows))
+  }
+
+  /**
+   * Delete all events before a sequenceNr for the given persistence id.
+   * @param persistenceId the persistence id to delete for
+   * @param toSequenceNr sequence nr (inclusive) to delete up to
+   */
+  def deleteTo(persistenceId: String, toSequenceNr: Long): Future[Done] = {
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:162]: `deleteEventsTo` to make it clear that it's not deleting snapshots?

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala [hunk 6]
```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
...
+    (journal ? CassandraJournal.DeleteEventsTo(persistenceId, toSequenceNr)).mapTo[Done]
+  }
+
+  /**
+   * Deletes all but the last N snapshots and deletes all events before this snapshot
+   * Does not delete from the tag_views table
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:168]: Perhaps additional caveat about consequences for this? Events by tag would fail if you run it and see a tag for a message that is not there?

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala [hunk 7]
```diff
@@ -220,6 +220,10 @@ import akka.stream.scaladsl.Source
         else result.map(_ => Done)
       result2.pipeTo(sender())
 
+    case DeleteEventsTo(persistenceId, toSequenceNr) =>
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:223]: what's the difference of this and the existing `akka.persistence.DeleteMessagesTo`?

**Review by johanandren — APPROVED** [review]
LGTM, some smaller nitpicky stuff mostly

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStatements.scala [hunk 8]
```diff
@@ -79,6 +86,19 @@ import akka.persistence.cassandra.FutureDone
...
+    persistence_id = ? 
+    ORDER BY sequence_nr DESC
+    LIMIT ?
+    """
+
+  def selectAllSnapshotMeta =
+    s"""SELECT sequence_nr, timestamp FROM ${tableName} WHERE
+    persistence_id = ? 
+    ORDER BY sequence_nr DESC
+    """
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStatements.scala:100]: can we use the existing `selectSnapshotMetadata` instead of adding two more?
the table is ordered by sequence_nr DESC

File: core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala [hunk 9]
```diff
@@ -192,6 +192,16 @@ abstract class CassandraSpec(
             println(s"""Row:${row.getLong("partition_nr")}, ${row.getString("persistence_id")}, ${row.getLong(
               "sequence_nr")}""")
           })
+
+        println("snapshots")
```
Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala:196]: cleanup

File: core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala (hunk 1 above)
Comment (chbatey) [reply core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala:78]: Yeah as we have log capturing on all the tests

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala [hunk 10]
```diff
@@ -42,18 +47,169 @@ final class Cleanup(systemProvider: ClassicActorSystemProvider, settings: Cleanu
...
+      snapshotsToKeep: Int,
+      keepAfter: Long): Future[Option[SnapshotMetadata]] = {
+    require(snapshotsToKeep >= 1, "must keep at least one snapshot")
+    require(keepAfter >= 0, "keepAfter must be greater than 0")
+    selectAllSnapshotMetaPs
+      .flatMap { ps =>
+        val allRows: Source[Row, NotUsed] = session.select(ps.bind(persistenceId))
+        @volatile var count = 0
+        allRows
+          .takeWhile { row =>
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:126]: could we skip the count hack with:
```
allRows
  .take(snapshotsToKeep)
  .takeWhile { row =>
```

?

**Review by patriknw — COMMENTED** [review]
looking good, a few questions

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala (hunk 3 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:136]: Much better. My brain was not working end of day yesterday

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala (hunk 4 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:164]: Unfortunately not as the range delete doesn't know what it deletes. Just adds a tombstone that reads can see

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala (hunk 5 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:162]: 👍

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 7 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:223]: That one doesn't respond to the sender() but the persistentActor in the message

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStatements.scala (hunk 8 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStatements.scala:100]: I think we need them. 
The first as we want to decide the LIMIT per query rather than fixed. 
The second as we don't want to provide a sequence nr. We could use 0 and Long.MAX but seems better to let C* know we're not filtering on sequence_nr

File: core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala (hunk 7 above)
Comment (patriknw) [reply core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala:223]: ok, but you could use the other ask to pass in the "ask sender" in the message. You find it in ExplicitAskSupport

File: core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala (hunk 9 above)
Comment (chbatey) [reply core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala:196]: ?

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala (hunk 10 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:126]: No as we may wan to keep more than snapshotsToKeep if they are all newer than the toKeep time

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala (hunk 6 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:168]: I don't think it'll fail, it doesn't reconcile that way around. Only messages to tag_views table. I will add warnings tho

File: core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala [hunk 11]
```diff
@@ -345,13 +291,17 @@ import akka.stream.alpakka.cassandra.scaladsl.{ CassandraSession, CassandraSessi
   sealed trait CleanupCommand
   final case class DeleteAllsnapshots(persistenceId: String) extends CleanupCommand
 
-  private case class Serialized(serialized: ByteBuffer, serManifest: String, serId: Int, meta: Option[SerializedMeta])
+  case class Serialized(serialized: ByteBuffer, serManifest: String, serId: Int, meta: Option[SerializedMeta])
 
-  private case class SerializedMeta(serialized: ByteBuffer, serManifest: String, serId: Int)
+  case class SerializedMeta(serialized: ByteBuffer, serManifest: String, serId: Int)
```
[the remark this answers is not in the document]
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala:296]: They're all in a private[akka] object

File: core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala (hunk 9 above)
Comment (chbatey) [reply core/src/test/scala/akka/persistence/cassandra/CassandraSpec.scala:196]: Ah the println? All the logging here is done with println

File: core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala (hunk 2 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/cleanup/Cleanup.scala:118]: Yah i struggled to describe it. 

Just having a timestamp makes it easy to understand "delete all snapshots older than X" but is an API that will lead to ppl deleting all their snapshots!

**Review by johanandren — APPROVED** [review]

File: core/src/main/resources/reference.conf [hunk 12]
```diff
@@ -505,6 +505,10 @@ akka.persistence.cassandra {
     # Log progress after this number of delete operations. Can be set to 1 to log
     # progress of each operation.
     log-progress-every = 100
+
+    # By default no deletes are executed and are instead logged at INFO. Set this to true
+    # to actually do the deletes
+    dry-run = true
```
Comment (patriknw) [inline core/src/main/resources/reference.conf:511]: mention existence of this at more places?

File: docs/src/main/paradox/cleanup.md [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+* Delete all snapshots and events for a persistence id keeping the latest N snapshots and all the events after them. 
+
+The cleanup tool can be combined with the @ref[query plugin](./read-journal.md) which has a query to get all persistence ids.
+
+Scala
+: @@snip [snapshot-keyspace](/docs/src/test/scala/doc/cleanup/CleanupDocExample.scala) { #cleanup } 
+
+Java
+: @@snip [snapshot-keyspace](/docs/src/test/java/jdoc/cleanup/CleanupDocExample.java) { #cleanup } 
+
```
Comment (patriknw) [inline docs/src/main/paradox/cleanup.md:29]: mention dry-run somewhere here

File: docs/src/main/paradox/cleanup.md [hunk 14]
```diff
@@ -0,0 +1,30 @@
+# Database Cleanup
```
Comment (patriknw) [inline docs/src/main/paradox/cleanup.md:1]: Good with a separate page. Also link to this from journal.md "Delete all events", snapshot.md "Delete all snapshots" and events-by-tag.md "Cleanup of tag_views table"

**Review by patriknw — APPROVED** [review]
LGTM, after some small doc additions

File: docs/src/main/paradox/cleanup.md (hunk 14 above)
Comment (chbatey) [reply docs/src/main/paradox/cleanup.md:1]: 👍
